File: Problems/33. Search in Rotated Sorted Array.cpp
#include<vector>
#include<iostream>


using namespace std;


class Solution {
    public:

        int binarySearch(vector<int>& nums, int target, int left, int right) {

            if(right>=left){

                int mid = left + (right - left) /2;

                cout << endl;
                cout << "mid: " << mid << " val: " << nums[mid] << " target: " << target << endl;
                cout << "Right: " << right << " And Left: " << left << endl;
                cout << endl;


                if(nums[mid]==target){
                    return mid;
                }


                if(target>nums[mid]){

                    if(target > nums[mid]){
                        return binarySearch(nums,target,mid+1,right);
                    }
                    else{
                        return binarySearch(nums,target,left,mid-1);
                    }

                    // if(target > nums[right]){
                    //     return binarySearch(nums,target,left,mid-1);
                    // }
                    // else{   
                    //     return binarySearch(nums,target,mid+1,right);
                    // }
                }
                if(target<nums[mid]){

                    if(target < nums[left]){
                        return binarySearch(nums,target,mid+1,right);
                    }
                    else{
                        return binarySearch(nums,target,left,mid-1);
                    }
                }


            }

            return -1;
        }
        

        int search(vector<int>& nums, int target) {
            return binarySearch(nums,target,0,nums.size()-1);
        }
};


void printVector(vector<int> nums){
    cout << " ";
    for(int i=0;i<nums.size();i++){
        cout << " " << nums[i] << " ";
    }
    cout << endl;
}

int main(){

    Solution solution;

    vector<int> input = {4,5,6,7,8,1,2,3};


    printVector(input);


    int answer = solution.search(input,8);

    cout << "Answer: " << answer << endl;

}

File: 33. Search in Rotated Sorted Array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        

        def bs(l,r):

            if l <= r:

                mid = l + (r - l) //2

                if nums[mid] == target: return mid

                # Check if we are in left part of the array or right
                # We are in left
                if nums[l] <= nums[mid]:
                    # Target is greater then max value on left subarray
                    if target > nums[mid]: return bs(mid+1,r)
                    # Target is less then least value on left subarray
                    elif target < nums[l]: return bs(mid+1,r) 
                    # Target is present in left subarray
                    else: return bs(l,mid-1)
                # Else we are in right sorted portion
                else:
                    # If target is greater than max number on subarray 
                    if target > nums[r]: return bs(l,mid-1)
                    #  Target is less then middle, so we just look at left portion of subarray
                    elif target < nums[mid]: return bs(l,mid-1)
                    # Target is present in right portion of subarray
                    else: return bs(mid+1,r)
                    
            return -1
        
        return bs(0,len(nums)-1)